SI-9390 Emit local defs that don't capture this as static

retronym opened the pull request:
This avoids unnecessary memory retention, and allows lambdas
that call the local methods to be serializable, regardless of
whether or not the enclosing class is serializable.

The second point is especially pressing, given that the enclosing
class for local methods defined in a used to be the (serializable)
anonymous function class, but as of Scala 2.12 will be the enclosing
class of the lambda.

This change is similar in spirit to SI-9408 / 93bee55e.

retronym committed:
SI-9390 Emit local defs that don't capture this as static

This avoids unnecessary memory retention, and allows lambdas
that call the local methods to be serializable, regardless of
whether or not the enclosing class is serializable.

The second point is especially pressing, given that the enclosing
class for local methods defined in a used to be the (serializable)
anonymous function class, but as of Scala 2.12 will be the enclosing
class of the lambda.

This change is similar in spirit to SI-9408 / 93bee55e.

retronym commented:
See https://gist.github.com/JoshRosen/8aacdee0162da430868e7f73247d45d8 for an analysis of this in the context of Spark. /cc @joshrosen

retronym commented:
Review by @adriaanm | @lrytz

retronym commented:
Bytecode diff from 2.12.0-M4 for a simple example:

```diff
 scala> :paste -raw
 // Entering paste mode (ctrl-D to finish)
 
 class StaticLift { def foo = { def bar(x: Any) = true; bar _ } }
 
 // Exiting paste mode, now interpreting.
 
 
 scala> :javap -c -private StaticLift
 Compiled from "<pastie>"
 public class StaticLift {
   public scala.Function1<java.lang.Object, java.lang.Object> foo();
     Code:
-       0: aload_0
-       1: invokedynamic #37,  0             // InvokeDynamic #0:apply:(LStaticLift;)Lscala/Function1;
-       6: areturn
+       0: invokedynamic #36,  0             // InvokeDynamic #0:apply:()Lscala/Function1;
+       5: areturn
 
-  private final boolean bar$1(java.lang.Object);
+  private static final boolean bar$1(java.lang.Object);
     Code:
        0: iconst_1
        1: ireturn
 
-  public final boolean StaticLift$$$anonfun$1(java.lang.Object);
+  public static final boolean StaticLift$$$anonfun$1(java.lang.Object);
     Code:
        0: aload_0
-       1: aload_1
-       2: invokespecial #46                 // Method bar$1:(Ljava/lang/Object;)Z
-       5: ireturn
+       1: invokestatic  #45                 // Method bar$1:(Ljava/lang/Object;)Z
+       4: ireturn
 
   public StaticLift();
     Code:
        0: aload_0
-       1: invokespecial #50                 // Method java/lang/Object."<init>":()V
+       1: invokespecial #49                 // Method java/lang/Object."<init>":()V
        4: return
 
-  public final java.lang.Object StaticLift$$$anonfun$1$adapted(java.lang.Object);
+  public static final java.lang.Object StaticLift$$$anonfun$1$adapted(java.lang.Object);
     Code:
        0: aload_0
-       1: aload_1
-       2: invokevirtual #52                 // Method StaticLift$$$anonfun$1:(Ljava/lang/Object;)Z
-       5: invokestatic  #58                 // Method scala/runtime/BoxesRunTime.boxToBoolean:(Z)Ljava/lang/Boolean;
-       8: areturn
+       1: invokestatic  #51                 // Method StaticLift$$$anonfun$1:(Ljava/lang/Object;)Z
+       4: invokestatic  #57                 // Method scala/runtime/BoxesRunTime.boxToBoolean:(Z)Ljava/lang/Boolean;
+       7: areturn
 
   private static java.lang.Object $deserializeLambda$(java.lang.invoke.SerializedLambda);
     Code:
        0: aload_0
-       1: invokedynamic #70,  0             // InvokeDynamic #1:lambdaDeserialize:(Ljava/lang/invoke/SerializedLambda;)Ljava/lang/Object;
+       1: invokedynamic #69,  0             // InvokeDynamic #1:lambdaDeserialize:(Ljava/lang/invoke/SerializedLambda;)Ljava/lang/Object;
        6: areturn
-}
+}
\ No newline at end of file

```

lrytz commented on test/junit/scala/tools/nsc/backend/jvm/opt/InlineWarningTest.scala:
```diff
@@ -141,7 +141,7 @@ class InlineWarningTest extends ClearAfterClass {
     val warn =
       """M::f()I is annotated @inline but could not be inlined:
-        |The callee M::f()I contains the instruction INVOKESPECIAL M.nested$1 ()I
+        |The callee M::f()I contains the instruction INVOKESTATIC M.nested$1 ()I
```
a nice thing about static methods is that we can in principle make them public for the purpose of the optimizer and don't need to name-mangle

lrytz commented on src/compiler/scala/tools/nsc/transform/Delambdafy.scala:
```diff
@@ -336,6 +352,9 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
         // but we do need to make note of the lifted body method in case it refers to 'this'
         if (currentMethod.exists) liftedMethodReferences(currentMethod) += targetMethod(fun)
+      case Apply(sel @ Select(This(_), _), args) if currentMethod.exists && sel.symbol.isLiftedMethod =>
+        if (currentMethod.exists) liftedMethodReferences(currentMethod) += sel.symbol
```
this condition looks redundant

JoshRosen commented:
Good news: I tested out this PR as of https://github.com/scala/scala/pull/5099/commits/e6a566e0b7a1c85c796bda44c809c6ea38cffd57 and it appears to have fixed the serializability issues with local `def`s that I found in Spark.

Bad news: in more local testing, I discovered a related problem affecting local classes. For example:

```scala
class MyClass {
  Seq().map { x: Any => new Object {} }
}
```

In this example, the `map`'s closure will become an instance method of `MyClass` (rather than a static method) because the anonymous class created by `new Object {}` takes an instance of `MyClass` in its constructor.

This sounds like [SI-1419](https://issues.scala-lang.org/browse/SI-1419); would it be possible to target that issue for 2.12?

retronym commented:
@JoshRosen That's a tricky one. As of [SI-9408](https://issues.scala-lang.org/browse/SI-9408) / 93bee55, the constructor of the anonymous class no longer stores retains a reference to the enclosing class, but its constructor does still include this:

```
class MyClass {
  def test = {
    { (x: Any) => new Object {} }
  }
}
```

```
  final class anon$1 extends Object {
    def <init>($outer: MyClass): <$anon: Object> = {
      anon$1.super.<init>();
      // $outer is not stored in a field
      ()
    }
  }
```

The transform responsible for this keeps the constructor signature intact because in general, another compilation unit might be calling the constructor of an inner class, and we need to have an agreed constructor signature given that client and class might be separately compiled.

However, for local classes like in your example, we could in fact totally remove `$outer` from the constructor signature (we'd need to rewrite calls to the constructor in the same method, of course).

retronym committed:
SI-9390 Avoid needless outer capture with local classes

An existing optimization in `Constructors` elides the outer
field in member and local classes, if the class doesn't use
the outer reference. (Member classes also need to be final,
which is a secret handshake to say we're also happy to weaken
prefix matching in the pattern matcher.)

That optimization leaves the constructor signature as is: the
constructor still accepts the outer instance, but does not store
it. For member classes, this means that we can separately compile
code that calls the constructor.

Local classes need not be hampered by this constraint, we could
remove the outer instance from the constructor call too.

Why would we want to do this?

Let's look at the case before and after this commit.

Before:
```
class C extends Object {
  def foo(): Function1 = $anonfun();
  final <static> <artifact> def $anonfun$foo$1($this: C, x: Object): Object = new <$anon: Object>($this);
  def <init>(): C = {
    C.super.<init>();
    ()
  }
};
final class anon$1 extends Object {
  def <init>($outer: C): <$anon: Object> = {
    anon$1.super.<init>();
    ()
  }
}
```

After:

```
class C extends Object {
  def foo(): Function1 = $anonfun();
  final <static> <artifact> def $anonfun$foo$1(x: Object): Object = new <$anon: Object>(null);
  def <init>(): C = {
    C.super.<init>();
    ()
  }
};
final class anon$1 extends Object {
  def <init>($outer: C): <$anon: Object> = {
    anon$1.super.<init>();
    ()
  }
}

```

However, the status quo means that a lambda that
This in turn makes lambdas that refer to such classes serializable
even when the outer class is not itself serialiable.

I have not attempted to extend this to calls to secondary constructors.

retronym commented:
> However, for local classes like in your example, we could in fact totally remove $outer from the constructor signature (we'd need to rewrite calls to the constructor in the same method, of course). This won't be trivial to implement, but might be worthwhile.

Here's a first cut at this change:
https://github.com/retronym/scala/compare/ticket/9390...retronym:ticket/9390-2?expand=1

lrytz commented:
LGTM!

retronym commented:
Let's run this one by @adriaanm next week before merging.

adriaanm commented:
LGTM -- should we look into unmangling names of members that end up being static, to keep them shorter?

adriaanm commented:
Sorry about the delay here -- unfortunately this will need a rebase.

retronym committed:
Lambda impl methods static and more stably named

The body of lambdas is compiled into a synthetic method
in the enclosing class. Previously, this method was a public
virtual method named `fully$qualified$Class$$anonfun$n`.
For lambdas that didn't capture a `this` reference, a static
method was used.

This commit changes two aspects.

Firstly, all lambda impl methods are now emitted static.
An extra parameter is added to those that require a this
reference.

This is an improvement as it:

  - allows, shorter, more readable names for the lambda impl method
  - avoids pollution of the vtable of the class. Note that javac uses
    private instance methods, rather than public static methods. If
    we followed its lead, we would be unable to support important use
    cases in our inliner

Secondly, the name of the enclosing method has been included in
the name of the lambda impl method to improve debuggability and
to improve serialization compatibility. The serialization improvement
comes from the way that fresh names for the impl methods are
allocated: adding or removing lambdas in methods not named "foo" won't
change the numbering of the `anonfun$foo$n` impl methods from methods
named "foo". This is in line with user expectations about anonymous
class and lambda serialization stability. Brian Goetz has described
this tricky area well in:

  http://cr.openjdk.java.net/~briangoetz/eg-attachments/lambda-serialization.html

This commit doesn't go as far a Javac, we don't use the hash of the
lambda type info, param names, etc to map to a lambda impl method name.
As such, we are more prone to the type-1 and -2 failures described there.
However, our Scala 2.11.8 has similar characteristics, so we aren't going
backwards.

Special case in the naming: Use "new" rather than "<init>" for constructor enclosed
lambdas, as javac does.

I have also changed the way that "delambdafy target" methods are identifed.
Rather than relying on the naming convention, I have switched to using a
symbol attachment. The assumption is that we only need to identify them
from within the same compilation unit.

This means we can distinguish impl metbods for expanded functions
(ones called from an `apply` method of an ahead-of-time expanded
anonfun class), from those that truly end up as targets for lambda
metafactory. Only the latter are translated to static methods in
this patch.

lrytz committed:
Treat self parameter as non-null in the optimizer

retronym commented:
@lrytz @adriaanm I've rebased this on top of another open PR that also touches `Delambdafy`.

Additionally, I have added support for avoiding capturing the enclosing class for lambdas that only need to instantiate local classes elided outer fields.

retronym commented:
/rebuild

retronym commented:
Looks like that null is getting used somewhere.

```
[quick.partest-extras] Exception in thread "main" java.lang.NullPointerException
[quick.partest-extras] 	at scala.collection.LinearSeqLike$$anon$1.hasNext(LinearSeqLike.scala:43)
[quick.partest-extras] 	at scala.collection.TraversableOnce.collectFirst(TraversableOnce.scala:144)
[quick.partest-extras] 	at scala.tools.util.PathResolver$Environment$.searchForBootClasspath(PathRes
```

lrytz commented:
```
trait LinearSeqLike .. { self =>
  def iterator: Iterator = new AbstractIterator {
    var these = self
    def hasNext = !these.isEmpty
  }
}
```

The anonymous class constructor has a parameter `$outer: LinearSeqLike` which is only used in the constructor (stored in field `these`), so no field for `$outer` is needed; but its value is still required..

retronym commented:
@lrytz Thanks for investigating. Hopefully it will be as easy as adding a check to `Constructors` that the outer param is not referenced in any of the constructors, other than to forward to the primary constructor.

adriaanm commented:
Adding the WIP label since scabot will insist on adding "reviewed" based on earlier comments.

retronym commented:
Still trying:

```
[quick.library] Exception in thread "main" java.lang.NullPointerException
[quick.library] 	at scala.tools.nsc.transform.Constructors$TemplateTransformer.scala$tools$nsc$transform$Constructors$TemplateTransformer$$$anonfun$34(Constructors.scala:718)
[quick.library] 	at 
```

retronym committed:
Avoid tree sharing with substituteThis

The underlying transformer has a by-name parameter for the
to provide the `to` tree, but this was strict in the layers
of API above.

Tree sharing is frowned upon in general as it leads to cross
talk when, e.g., the erasure typechecker mutates the
`tpe` field of the shared tree in different context.

retronym commented:
@adriaanm @lrytz The last commit here needs review now.

lrytz commented:
@adriaanm fyi, the first three commits here are the same as in https://github.com/scala/scala/pull/5157

lrytz commented:
the last commit message seems incomplete:

```
However, the status quo means that a lambda that
This in turn makes lambdas that refer to such classes serializable
even when the outer class is not.
```

lrytz commented on src/compiler/scala/tools/nsc/transform/Delambdafy.scala:
```diff
@@ -368,6 +371,9 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
       case Apply(sel @ Select(This(_), _), args) if sel.symbol.isLiftedMethod =>
         if (currentMethod.exists) liftedMethodReferences(currentMethod) += sel.symbol
         super.traverseTrees(args)
+      case Apply(fun, outer :: rest) if fun.symbol.isConstructor && treeInfo.isQualifierSafeToElide(outer) && fun.symbol.hasAttachment[OuterArgCanBeElided.type] =>
```
maybe extract this condition into a local method, as it's used in two places

retronym replied:
done

lrytz commented:
looks good otherwise!

adriaanm commented:
ping @retronym -- lets apply a quick round of polish based on Lukas's feedback and get this merged today

lrytz commented:
note that merging this will also merge the commits it shares with #5157

adriaanm commented:
Ok, I'm going to let you coordinate these two then.

lrytz commented:
LGTM
